### PR TITLE
Redo serialize.lua

### LINKED ISSUE #8719: minetest.serialize produces bad output for some tables with circular references
##### Minetest version

```
Minetest 5.0.1
```

##### OS / Hardware

Operating system: Windows 10 Home
CPU: Intel i7-3770K @ 3.5GHz, 8.00 GB ram

##### Summary

In the course of attempting to persist some tables between runs I've discovered that minetest.serialize() produces broken output for some (but not all) tables with circular references. The resulting output crashes with a nil dereference error when attempting to loadstring() it back into a table again; I'll detail a reliable example in the "steps to reproduce" section below.

This doesn't occur in all cases, I've successfully serialized and deserialized tables with circular references before. It appears that a particular pattern of nested circular references are required to trigger this bug.

##### Steps to reproduce

The following code can be used as a standalone init.lua mod file to illustrate the problem:

```
-- Initial non-circular-reference-containing data structure:
local data = {
	player_accounts = {
		player1 = {
			log = {
				{}
			},
		},
		player2 = {
			log = {},
		},
	},
}

-- Adding the circular references. The two player accounts have a log entry that
-- they share a reference to, and the log entry references both player accounts.
data.player_accounts.player1.log[1].seller = data.player_accounts.player1
data.player_accounts.player1.log[1].purchaser = data.player_accounts.player2
data.player_accounts.player2.log[1] = data.player_accounts.player1.log[1]

local serialized = minetest.serialize(data)
minetest.debug("Source data:")
minetest.debug(dump(data))
minetest.debug("Serialized data:")
minetest.debug(serialized)
-- minetest.serialize outputs Lua code that, when executed, builds and returns
-- the original data structure. minetest.deserialize() simply returns nil on
-- failure, so I manually loadstring() and execute the result here to make
-- the error explicit:
local loadedstring = loadstring(serialized)
local data2 = loadedstring() -- This step crashes
minetest.debug("Loadstring data:")
minetest.debug(dump(data2))
```

This is the serialized version of the ``data`` structure that minetest.serialize produces, with added comments showing where it's gone wrong:

```
local _ = {}
_[1] = {}
_[2] = {["log"] = _[1]}
_[3] = {}
_[4] = {["log"] = _[3]}
-- There should be "_[5] = {}" at this line, but there is not.
_[3][1] = _[5] -- index 5 is nil at this point so this line does nothing
_[5]["purchaser"] = _[2] -- this line crashes with an attempt to index nil.
_[5]["seller"] = _[4]
_[1][1] = _[5]
return {["player_accounts"] = {["player2"] = _[2], ["player1"] = _[4]}}
```

### LINKED ISSUE #11899: minetest.deserialize fails for serialized data containing NaN or infinities
##### Minetest version
Any

##### Summary
The way `minetest.serialize` currently formats numbers, NaN will be turned into `nan` or `-nan`, depending on the sign; infinities will be turned into `-inf` or `inf`. The environment `minetest.deserialize` provides however contains neither `nan` nor `inf` variables. Both will therefore evaluate to `nil`. If they are negated, this will lead to a crash (`attempt to negate a nil value`). It will also lead to a crash if they are used as a table index; this isn't an issue for NaN, as it can't occur as a table index anyways, but tables containing positive infinity as a key will lead to an `table index is nil`. The `pcall` will catch this, but the returned object will be `nil` if an error has occurred.

In a nutshell:

* Negative NaN or negative infinity anywhere within the data as well as positive infinity as a table key will cause deserialization to fail;
* Positive infinity or positive NaN elsewhere will be turned into `nil`.

##### Steps to reproduce
The below assertions will all fail:
```lua
assert(minetest.deserialize(minetest.serialize(0/0)))
assert(minetest.deserialize(minetest.serialize(math.huge)))
assert(minetest.deserialize(minetest.serialize({[math.huge] = 0})))
```

### PR DESCRIPTION
**Pull request by appgurueu**
Alternative to #11380. Not yet thoroughly tested, but passes the (few) unit tests provided, as well as my [own tests](https://github.com/appgurueu/modlib/blob/master/test.lua#L169-L198). Closes #8719.

Features:

* Support for arbitrary references, including self-referencing
* Short output, references "long" strings as a bonus
* Around the same speed, potentially slower if long, short keys are present
* Properly works with NaN and inf (fixes #11899)

<details>
  <summary>Benchmarks</summary>
 
LuaJIT. Write speeds, Luon is the name of the modlib module, Bluon is a binary format, Minetest is the old `minetest.serialize`, Minetest (JSON) is `minetest.write_json`. The times shown are what 100 serializations take. Data from 
https://github.com/miloyip/nativejson-benchmark/tree/master/data.
```
OBJECT: canada.json
Luon    13.800655s
Bluon   21.086339s
Minetest        16.716192s
Minetest (JSON) 41.147817s
OBJECT: citm_catalog.json
Luon    3.456392s
Bluon   3.230237s
Minetest        3.811432s
Minetest (JSON) 5.631833s
OBJECT: twitter.json
Luon    1.186978s
Bluon   0.79831700000001s
Minetest        1.015294s
Minetest (JSON) 1.718552s
```
  
</details>

("reopened" #11425)

### CODE REVIEW

File: builtin/common/serialize.lua [hunk 1]
```diff
@@ -1,205 +1,190 @@
...
+            references[object] = ref
+            if type_ == "table" then
+                to_fill[object] = ref
+            end
+            increment_reference(1)
+        end
+    end
+    local function is_short_key(key)
+        return not references[key] and type(key) == "string" and string_match(key, "^[%a_][%a%d_]*$")
+    end
```
Comment (SmallJoker) [inline builtin/common/serialize.lua:75]: Misleading function name. Should be something similar to `is_short_key_notation_possible`

File: builtin/common/serialize.lua [hunk 2]
```diff
@@ -1,205 +1,190 @@
...
-				else
-					mark_multiple_occurences(k)
-					mark_multiple_occurences(v)
-				end
-			end
-			nested[x] = nil
-		end
+-- Recursively counts occurences of objects (non-primitives including strings) in a table.
+local function count_objects(value)
+    local counts = {}
```
Comment (SmallJoker) [inline builtin/common/serialize.lua:13]: Please use tabulators

File: builtin/common/serialize.lua [hunk 3]
```diff
@@ -1,205 +1,190 @@
...
+	if value == nil then
+		-- Early return for nil
+		return counts
 	end
+    local function count_values(val)
+		local type_ = type(val)
+		if type_ == "boolean" or type_ == "number" then return end
+        local count = counts[val]
```
Comment (SmallJoker) [inline builtin/common/serialize.lua:21]: Misleading indent extent

File: builtin/common/serialize.lua [hunk 4]
```diff
@@ -1,205 +1,190 @@
...
+    local function increment_reference(place)
+        if not reference[place] then
+            reference[place] = "B"
+        elseif reference[place] == "Z" then
+            reference[place] = "A"
+            return increment_reference(place + 1)
+        else
+            reference[place] = succ[reference[place]]
+        end
+    end
```
Comment (SmallJoker) [inline builtin/common/serialize.lua:55]: Are references used to shorten the output? Needs a comment in-line to describe what the purpose is, also that it's limited from A-Z.

File: builtin/common/serialize.lua [hunk 5]
```diff
@@ -1,205 +1,190 @@
...
+            local first = true
+            write"{"
+            local len = #value
+            for i = 1, len do
+                if not first then write";" end
+                dump(value[i])
+                first = false
+            end
+            for k, v in next, value do
+                if type(k) ~= "number" or k % 1 ~= 0 or k < 1 or k > len then
```
Comment (SmallJoker) [inline builtin/common/serialize.lua:110]: Needs a description what this logic is supposed to do. Why `k % 1 ~= 0`?

File: builtin/common/serialize.lua [hunk 6]
```diff
@@ -1,205 +1,190 @@
...
+                        write(k)
+                    else
+                        write"["
+                        dump(k)
+                        write"]"
+                    end
+                    write"="
+                    dump(v)
+                    first = false
+                end
```
Comment (SmallJoker) [inline builtin/common/serialize.lua:122]: Else?

File: builtin/common/serialize.lua [hunk 7]
```diff
@@ -1,205 +1,190 @@
...
+        end
+    end
+    local references = {}
+    local to_fill = {}
+    for object, count in pairs(count_objects(value)) do
+        local type_ = type(object)
+        if count >= 2 and (type_ ~= "string" or #reference + 2 < #object) then
+            local ref = table_concat(reference)
+            write(ref)
+            write"="
```
Comment (SmallJoker) [inline builtin/common/serialize.lua:63]: Inconsistent use of parentheses. Whereas this is a valid notation, it does make the code less readable.

File: builtin/common/serialize.lua [hunk 8]
```diff
@@ -1,205 +1,190 @@
...
+    return counts
+end
 
-	local dumped     = {}  -- object->varname set
-	local local_defs = {}  -- Dumped local definitions as source code lines
-
-	-- Mutually recursive local functions:
-	local dump_val, dump_or_ref_val
+-- Build a table with the succeeding character from A-Z
+local succ = {}
```
Comment (SmallJoker) [inline builtin/common/serialize.lua:35]: Not intuitive variable name. Proposal: `next_reference_character`
Needs a comment what this is used for, and not what it does. Letter is already obvious.

File: builtin/common/serialize.lua [hunk 9]
```diff
@@ -1,205 +1,190 @@
...
+        if ref then
+            return write(ref)
+        elseif type_ == "string" then
+            return write(quote(value))
+        elseif type_ == "table" then
+            local first = true
+            write"{"
+            local len = #value
+            for i = 1, len do
+                if not first then write";" end
```
Comment (SmallJoker) [inline builtin/common/serialize.lua:105]: Why not terminate it after each value, rather than checking for first occurrences?

**Review by SmallJoker — COMMENTED** [review]
Overall needs more comments to describe what each step does for easier maintenance and better understanding.
Erroneous code style
See comments

File: builtin/common/serialize.lua (hunk 9 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:105]: > Why not terminate it after each value, rather than checking for first occurrences?

Because I don't like trailing semicolons/commas. But it is an option. Either way, performance impact is pretty negligible.

File: builtin/common/serialize.lua (hunk 5 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:110]: This logic checks for non list keys.

File: builtin/common/serialize.lua (hunk 6 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:122]: No need for an else, the previous loop already wrote all list keys.

File: builtin/common/serialize.lua (hunk 1 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:75]: > Misleading function name. Should be something similar to `is_short_key_notation_possible`

How about `use_short_key`?

File: builtin/common/serialize.lua (hunk 2 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:13]: Will do

File: builtin/common/serialize.lua (hunk 3 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:21]: Will fix.

File: builtin/common/serialize.lua (hunk 4 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:55]: References are necessary for serializing circular structures (or structures in which a table occurs multiple times). They are additionally used to shorten the output, yes, by referencing strings.

File: builtin/common/serialize.lua (hunk 7 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:63]: Will adhere

File: builtin/common/serialize.lua (hunk 8 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:35]: > Not intuitive variable name. Proposal: `next_reference_character`
> Needs a comment what this is used for, and not what it does. Letter is already obvious.

I disagree. See Haskell:

```
$ ghci
GHCi, version 8.6.5: http://www.haskell.org/ghc/  :? for help
Prelude> succ 'A'
'B'
```

**Comment by appgurueu** [comment]
This still fixes #8719

**Comment by rubenwardy** [comment]
This should be unit tested to cover compatibility and edge cases, and also benchmarked

See how vectors are unit tests in `builtin/common/tests/vector_spec.lua`

**Comment by appgurueu** [comment]
> This should be unit tested to cover compatibility and edge cases
> 
> See how vectors are unit tests in `builtin/common/tests/vector_spec.lua`

Minetest already has unit tests for serialization and this passes them, plus my modlib unit tests (as already said). I should definitely add FaceDeer's example to the unit tests though.

**Comment by rubenwardy** [comment]
> Minetest already has unit tests for serialization and this passes them

oh right, I wrote them. It would be good to add tests for previous shortcomings, such as "Support for arbitrary references, including self-referencing"

File: builtin/common/serialize.lua (hunk 8 above)
Comment (JosiahWI) [reply builtin/common/serialize.lua:38]: `succ` doesn't seem intuitive to me either. Especially for people who don't know Haskell, I think SmallJoker has a good suggestion.

File: builtin/common/serialize.lua [hunk 10]
```diff
@@ -1,205 +1,217 @@
...
-			local idx_dumped = {}
-			local np = nest_points[x]
-			for i, v in ipairs(x) do
-				if not np or not np[i] then
-					vals[#vals + 1] = dump_or_ref_val(v)
-				end
-				idx_dumped[i] = true
+	local function use_short_key(key)
+		return not references[key] and type(key) == "string" and string_match(key, "^[%a_][%a%d_]*$")
+	end
```
Comment (Desour) [inline builtin/common/serialize.lua:95]: Pls comment this func.
Ie. something like:
`-- Used to decide whether we can do "key=..." instead of "["..quote(key).."]=..."`

File: builtin/common/serialize.lua [hunk 11]
```diff
@@ -1,205 +1,217 @@
...
+			return write(ref)
+		end
+        if type_ == "string" then
+			return write(quote(value))
+		end
+        if type_ == "function" then
+			return write(dump_func(value))
+		end
+        if type_ == "table" then
+			local first = true
```
Comment (Desour) [inline builtin/common/serialize.lua:123]: It wouldn't really hurt to have a superficial ";" at the end of each table. And the code would be simpler and possibly faster.

**Review by Desour — COMMENTED** [review]
Should work. Didn't test though.
Some more comments and empty lines wouldn't hurt.
(If you want to further improve the performance, I'd suggest using luajit's profiler.)

File: builtin/common/serialize.lua (hunk 8 above)
Comment (appgurueu) [reply builtin/common/serialize.lua]: `succ_ref_char` it is.

File: builtin/common/serialize.lua (hunk 10 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:95]:
```lua
					if use_short_key(k) then
						write(k)
					else
						write("[")
						dump(k)
						write("]")
					end
```

Is quite self-explaining, isn't it?

File: builtin/common/serialize.lua [hunk 12]
```diff
@@ -1,205 +1,217 @@
...
+			local ref = table_concat(reference)
+			write(ref)
+			write("=")
+            if type_ == "table" then
+                write("{}")
+            elseif type_ == "function" then
+                write(dump_func(object))
+            elseif type_ == "string" then
+                write(quote(object))
+            end
```
[the remark this answers is not in the document]
Comment (appgurueu) [reply builtin/common/serialize.lua]: This will be caught during writing anyways. But moving this to the counting is probably preferrable.

File: builtin/common/serialize.lua [hunk 13]
```diff
@@ -1,205 +1,217 @@
...
+            -- Carriage
+			reference[place] = "A"
+			return increment_reference(place + 1)
+		else
+			reference[place] = succ[reference[place]]
 		end
-		local var = dumped[x]
-		if var then  -- Already referenced
-			return var
+	end
```
[the remark this answers is not in the document]
Comment (appgurueu) [reply builtin/common/serialize.lua:70]: Indeed, this serves the purpose of shortening. `_` might have a special meaning, `0-9` are tricky because you might get something like `0x` which is tokenized as a number. I could include `a-z`, but that might be irritating to human readers...

File: builtin/common/serialize.lua [hunk 14]
```diff
@@ -1,205 +1,217 @@
...
+local function count_objects(value)
+	local counts = {}
+	if value == nil then
+		-- Early return for nil
+		return counts
+	end
+	local function count_values(val)
+		local type_ = type(val)
+		if type_ == "boolean" or type_ == "number" then
+            return
```
[the remark this answers is not in the document]
Comment (appgurueu) [reply builtin/common/serialize.lua]: Done.

File: builtin/common/serialize.lua (hunk 10 above)
Comment (Desour) [reply builtin/common/serialize.lua:99]: Yes, but it comes much later in code. Many functions are self-explaining if you look up how they are used, but this looking up is work.

Comment (appgurueu) [reply builtin/common/serialize.lua:99]: Well, I guess a comment wouldn't hurt

File: builtin/common/serialize.lua (hunk 11 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:128]: Hmm. I'm torn on that one. I don't like removing a feature... Guess I'll make it customizable in modlib, but here it is indeed unneeded.

File: builtin/common/serialize.lua (hunk 13 above)
[the remark this answers is not in the document]
Comment (Desour) [reply builtin/common/serialize.lua:70]: If you use all possible ASCII characters for names, you have of course make to sure that it's still valid names (ie. no numbers or keywords). This can be done by starting all names with an `A`.

I agree that including `a-z` can irritate human readers, especially with `l` and `I`. I hadn't thought about human-readability.

For simplicity I'd probably go with `"A"..counter`. But I don't really care.

[the remark this answers is not in the document]
Comment (appgurueu) [reply builtin/common/serialize.lua:70]: Alright then.

**Comment by appgurueu** [comment]
Bump. 5.5 could use this.

**Comment by appgurueu** [comment]
It looks like I'm duplicating part of the `is_valid_identifier` implementation (which is used by `dump`) with my latest changes. I'll see how I can deduplicate this (most likely I'll juse expose `is_valid_identifier` and localize it for performance).

**Comment by Df458** [comment]
From a brief non-thorough test with [one of my mods](https://content.minetest.net/packages/Hugues%20Ross/cartographer/) that serializes a fair amount of data, this seems to work as expected. Notably, the resulting data was significantly smaller with this patch due to the way keys are handled. The code itself also seems fine, I had to slow down reading it but it's still understandable.

I'd call this good to merge with a couple caveats
- This could probably do with some "real" use-case testing with mods/games that rely on serialization over multiple play sessions
- I haven't checked code style yet, although I don't expect to find anything since SmallJoker did a pass for that already

I think the best time would be _directly after 5.5 is released_ though, the potential harm that a missed bug could do here is very high and this will give time for us to see more real-world behavior during the 5.6 release cycle.

**Comment by appgurueu** [comment]
> I think the best time would be _directly after 5.5 is released_ though, the potential harm that a missed bug could do here is very high and this will give time for us to see more real-world behavior during the 5.6 release cycle.

Agreed. Get 5.5 out first, it's already late. I'll see where I can find good real-world test cases.

**Comment by appgurueu** [comment]
The latest commit massively improves the coverage, testing thousands of primitives and tables through fuzzing. I had to write my own `assert_same` because [busted's deepcompare](https://github.com/Olivine-Labs/luassert/blob/e2ab0d218d7a63bbaa2fdebfa861c24a48451e9d/src/util.lua#L12) (1) only compares by depth, not by reference "mappability" and (2) simply didn't complete (poor performance, possibly exponential?).

**Comment by appgurueu** [comment]
The LuaJIT tests are currently failing due to what appears to be a LuaJIT 2.1.0-beta3 bug. Further investigation is needed.

File: builtin/common/tests/serialize_spec.lua [hunk 15]
```diff
@@ -6,44 +6,98 @@ _G.setfenv = require 'busted.compatibility'.setfenv
...
+		test_in.bar.baz[1] = test_in.foo.baz[1]
+
+		assert_preserves(test_in)
 	end)
 
 	it("strips functions in safe mode", function()
 		local test_in = {
 			func = function(a, b)
-				error("test")
+				return nil * nil -- would throw an error if called
```
Comment (sfan5) [inline builtin/common/tests/serialize_spec.lua:100]: ?

Comment (appgurueu) [reply builtin/common/tests/serialize_spec.lua:100]: LuaJIT was unable to `string.dump` the function before as it uses an upvalue (the environment containing `error`).

Comment (sfan5) [reply builtin/common/tests/serialize_spec.lua:100]: does `_G.error("test")` work or does busted also mess with that?

Comment (appgurueu) [reply builtin/common/tests/serialize_spec.lua:100]: > does `_G.error("test")` work or does busted also mess with that?

Would suffer from the same problem I suppose as `_G` still indexes the environment inherited from the `it` func.

**Comment by appgurueu** [comment]
It seems like this bug is fixed on the latest master of LuaJIT so I've changed the workflow to compile that rather than installing the outdated package from the Ubuntu repos.

File: builtin/common/tests/serialize_spec.lua (hunk 15 above)
Comment (appgurueu) [reply builtin/common/tests/serialize_spec.lua]: I've switched it to `setfenv` in https://github.com/minetest/minetest/pull/11427/commits/36c8e5a7f18d7254d9f8991cbc8cd5dd36c5f577

**Comment by Df458** [comment]
> It seems like this bug is fixed on the latest master of LuaJIT so I've changed the workflow to compile that rather than installing the outdated package from the Ubuntu repos.

I'm tired so maybe this is a stupid question, but--could that bug affect users of Minetest on Ubuntu if we merge this? And if so, general question for the peanut gallery: Should we assume that a user taking steps to pull in the latest version of Minetest on that platform (since iirc they don't push updates for it often) will also be the sure to pull in newer deps?

It's possible this is a non-issue, or that we're comfortable with the latter. I just figure it's worth throwing that out for discussion while I look this over.

**Comment by appgurueu** [comment]
> I'm tired so maybe this is a stupid question, but--could that bug affect users of Minetest on Ubuntu if we merge this? And if so, general question for the peanut gallery: Should we assume that a user taking steps to pull in the latest version of Minetest on that platform (since iirc they don't push updates for it often) will also be the sure to pull in newer deps?

This bug happened to me only when running tests using busted, never outside of the busted environment. It also doesn't occur in the serializer but rather in the test code: Fuzzing a primitive yields a C function for some reason. I've managed to isolate the bug to the following snippet of code:

**`luajitwat.lua`**:
```lua
describe("fuzz", function()
	local function charcodes(count)
		if count == 0 then return end
		return math.random(0, 0xFF), charcodes(count - 1)
	end
	local function str()
		return string.char(charcodes(math.random(0, 10)))
	end
	for _ = 1, 1e6 do
		local fuzzstr = str()
		if type(fuzzstr) == "function" then
			error(debug.getinfo(fuzzstr, "S").source)
		end
	end
end)
```

running this sometimes fails *despite Lua's random being deterministic*:

```
$ busted luajitwat.lua --lua=luajit
# successful run, no error
0 successes / 0 failures / 0 errors / 0 pending : 0.35087 seconds
$ busted luajitwat.lua --lua=luajit
✱
0 successes / 0 failures / 1 error / 0 pending : 0.00165 seconds
# failing run
Error → luajitwat.lua @ 1
fuzz
luajitwat.lua:12: =[C]
```

> It's possible this is a non-issue, or that we're comfortable with the latter. I just figure it's worth throwing that out for discussion while I look this over.

It should be a non-issue for the PR. LuaJIT being outdated overall is [an issue](https://github.com/LuaJIT/LuaJIT/issues/829) but this is unrelated to this PR (apart from the busted testing workflow, which I fixed).

**Review by Df458 — APPROVED** [review]

File: .github/workflows/lua.yml [hunk 16]
```diff
@@ -40,23 +40,33 @@ jobs:
...
+      - uses: actions/checkout@v3
+      - uses: leafo/gh-actions-lua@v9
+        with:
+          luaVersion: "5.1.5"
+      - uses: leafo/gh-actions-luarocks@v4
+
+      - name: Install LuaJIT
+        run: |
+          cd $HOME
+          git clone https://github.com/LuaJIT/LuaJIT/
```
Comment (sfan5) [inline .github/workflows/lua.yml:55]: I'd rather skip that one test on LuaJIT than this

also why does this entire file show as changed? whitespace?

File: builtin/common/serialize.lua [hunk 17]
```diff
@@ -1,205 +1,232 @@
...
-	local dumped     = {}  -- object->varname set
-	local local_defs = {}  -- Dumped local definitions as source code lines
+-- Build a "set" of Lua keywords. These can't be used as short key names.
+-- See https://www.lua.org/manual/5.1/manual.html#2.1
+local keywords = {}
+for _, keyword in pairs({
+	"and", "break", "do", "else", "elseif",
+	"end", "false", "for", "function", "if",
+	"in", "local", "nil", "not", "or",
+	"repeat", "return", "then", "true", "until", "while"
```
Comment (sfan5) [inline builtin/common/serialize.lua:47]: `goto` for luajit?
perhaps blacklisting `const` is worth it too to allow loading output in lua 5.3

File: builtin/common/serialize.lua [hunk 18]
```diff
@@ -1,205 +1,232 @@
...
+			-- Carriage
+			reference[place] = "A"
+			return increment_reference(place + 1)
+		else
+			reference[place] = succ_ref_char[reference[place]]
 		end
-		local var = dumped[x]
-		if var then  -- Already referenced
-			return var
+	end
```
Comment (sfan5) [inline builtin/common/serialize.lua:82]: I have the same thought here: why not just use `"r" .. n`?
The amount of saved space is minimal and keeping a table around that you have to increment and then concat (this was not immediately obvious to me btw) is IMO wasted work.

File: builtin/common/serialize.lua [hunk 19]
```diff
@@ -1,205 +1,232 @@
...
+		if type_ == "table" then
+			write("{")
+			-- First write list keys:
+			-- Don't use the table length #value here as it may horribly fail
+			-- for tables which use large integers as keys in the hash part;
+			-- stop at the first "hole" (nil value) instead by using ipairs
+			local len = 0
+			for i, v in ipairs(value) do
+				dump(v)
+				write(";")
```
Comment (sfan5) [inline builtin/common/serialize.lua:148]: should be a comma, I am surprised this is even valid.

File: builtin/common/serialize.lua [hunk 20]
```diff
@@ -1,205 +1,232 @@
...
+				len = i
 			end
-			for k, v in pairs(x) do
-				if (not np or not np[k]) and
-						not idx_dumped[k] then
-					vals[#vals + 1] = "["..dump_or_ref_val(k).."] = "
-						..dump_or_ref_val(v)
+			-- Now write map keys ([key] = value)
+			for k, v in next, value do
+				-- Check whether this is a non-list key (hash key)
```
Comment (sfan5) [inline builtin/common/serialize.lua:153]:
```suggestion
				-- We have written all non-float keys in [1, len] already
```

File: builtin/common/serialize.lua [hunk 21]
```diff
@@ -1,205 +1,232 @@
...
+					if use_short_key(k) then
+						write(k)
+					else
+						write("[")
+						dump(k)
+						write("]")
+					end
+					write("=")
+					dump(v)
+					write(";")
```
Comment (sfan5) [inline builtin/common/serialize.lua:164]: same

File: builtin/common/serialize.lua [hunk 22]
```diff
@@ -1,205 +1,232 @@
...
 		end
 	end
-
-	local function dump_nest_points()
-		for parent, vals in pairs(nest_points) do
-			for k, v in pairs(vals) do
-				local_defs[#local_defs + 1] = dump_or_ref_val(parent)
-					.."["..dump_or_ref_val(k).."] = "
-					..dump_or_ref_val(v)
+	for table, ref in pairs(to_fill) do
```
Comment (sfan5) [inline builtin/common/serialize.lua:171]: `-- Write the statements to fill circular tables`

File: builtin/common/serialize.lua [hunk 23]
```diff
@@ -1,205 +1,232 @@
...
-		return "local _ = {}\n"
-			..table.concat(local_defs, "\n")
-			.."\nreturn "..top_level
-	else
-		return "return "..top_level
-	end
+function core.serialize(value)
+	local rope = {}
+	serialize(value, function(text)
+		table_insert(rope, text)
```
Comment (sfan5) [inline builtin/common/serialize.lua:194]: I thought `rope[#rope + 1] = text` was faster here

File: builtin/common/serialize.lua [hunk 24]
```diff
@@ -1,205 +1,232 @@
...
+local function deserialize(func, safe)
+	-- math.huge is serialized to inf, 0/0 is serialized to -nan
+	local env = {inf = math_huge, nan = 0/0}
+	if safe then
+		env.loadstring = dummy_func
+	else
+		env.loadstring = function(str, ...)
+			if str:byte(1) == 0x1B then
+				return nil, "Bytecode prohibited"
+			end
```
Comment (sfan5) [inline builtin/common/serialize.lua:210]: handled by mod security and should be removed

File: builtin/common/serialize.lua [hunk 25]
```diff
@@ -1,205 +1,232 @@
...
-	-- pollute it with permanent references.
-	setfenv(f, {loadstring = safe and dummy_func or safe_loadstring})
-
-	local good, data = pcall(f)
-	if good then
-		return data
-	else
-		return nil, data
-	end
+	return deserialize(assert(loadstring(str)), safe)
```
Comment (sfan5) [inline builtin/common/serialize.lua:231]: why not inline this?

File: builtin/common/serialize.lua (hunk 17 above)
Comment (appgurueu) [reply builtin/common/serialize.lua]: Blacklisting `const` is not necessary as `const` is a variable attribute rather than a keyword; it must be used within angle brackets and Lua's tokenizer takes care of this properly. `{const = 1}` is valid in Lua 5.3.

Adding `goto` to the list definitely seems reasonable though. In fact I have added it on the [modlib equivalent](https://github.com/appgurueu/modlib/blob/b974113fa9781737a0f8e811ac83eecd5fa34092/text.lua#L180). Done.

File: builtin/common/serialize.lua (hunk 18 above)
Comment (appgurueu) [reply builtin/common/serialize.lua:83]: > The amount of saved space is minimal

It is a factor of `math.log(26)/math.log(10)` which is ca. `1.4` plus the one wasted character `r`. This probably makes for 2-3 characters per ref; on the other hand, not stringification of numbers might be faster. Can I at least use hex :>?

File: builtin/common/serialize.lua (hunk 24 above)
Comment (appgurueu) [reply builtin/common/serialize.lua]: done

File: builtin/common/serialize.lua (hunk 25 above)
Comment (appgurueu) [reply builtin/common/serialize.lua]: Inline what? This comment let me to restore backwards compat regarding serialized data that isn't valid Lua BTW.

File: builtin/common/serialize.lua (hunk 18 above)
Comment (sfan5) [reply builtin/common/serialize.lua:83]: Sure

File: builtin/common/serialize.lua (hunk 19 above)
Comment (sfan5) [reply builtin/common/serialize.lua]: I have never seen `;` used in tables so yes please

File: .github/workflows/lua.yml (hunk 16 above)
Comment (sfan5) [reply .github/workflows/lua.yml]: It would be to minimize external dependencies in CI but I supposed it's not that important.

File: builtin/common/serialize.lua (hunk 18 above)
Comment (appgurueu) [reply builtin/common/serialize.lua]: Done

**Comment by sfan5** [comment]
Can we cut down on trailing commas? 
grafik

**Comment by appgurueu** [comment]
> Can we cut down on trailing commas?

We can and we did. [Desour seems to prefer trailing commas though.](https://github.com/minetest/minetest/pull/11427#discussion_r710488614)

> Also these should probably be declared as locals: `r1="param1";r2="default:desert_stone";r3="default:desert_sand";r4="default:cactus"`

Locals are capped at 255. I don't want to increase the code complexity (or impose arbitrary limits) for a negligible performance benefit.

> Also² this breaks WorldEdit's [deserialize workaround on LuaJIT](https://github.com/Uberi/Minetest-WorldEdit/blob/c223ca4cec7cb17ddff30f3fd0a35a8c85ca42d5/worldedit/serialization.lua#L124-L143). It was bound to happen of course but it's not fun to figure out a fix.

Now I will worry about this.

**Comment by sfan5** [comment]
> Locals are capped at 255. I don't want to increase the code complexity (or impose arbitrary limits) for a negligible performance benefit.

Wasn't even thinking of (theoretical) performance benefits, but sure you can keep it like this.

File: builtin/common/serialize.lua (hunk 20 above)
Comment (sfan5) [reply builtin/common/serialize.lua:138]: what about this

Comment (appgurueu) [reply builtin/common/serialize.lua:138]: I thought I had committed this already, weird

**Review by sfan5 — APPROVED** [review]
Tested mainly with WorldEdit schematic serialization
